compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -3582,17 +3581,17 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
               case xtree => typedUnnamed(xtree)
 
           val unsimplifiedType = result.tpe
-          simplify(result, pt, locked)
-          result.tpe.stripTypeVar match
+          val result1 = simplify(result, pt, locked)
+          result1.tpe.stripTypeVar match
             case e: ErrorType if !unsimplifiedType.isErroneous => errorTree(xtree, e.msg, xtree.srcPos)
-            case _ => result
+            case _ => result1
         catch case ex: TypeError =>
           handleTypeError(ex)
      }
   }
 
   /** Interpolate and simplify the type of the given tree. */
-  protected def simplify(tree: Tree, pt: Type, locked: TypeVars)(using Context): tree.type =
+  protected def simplify(tree: Tree, pt: Type, locked: TypeVars)(using Context): Tree =
```

Comment:
Why is this part written? The original code seems fine to me because the type of the tree is updated in place.

Reply:
I must've missed it, when I was rewriting the implementation, it's not needed anymore. Changed.